### ISSUE TITLE
Correct total stress in HM coupling

### ISSUE THREAD

**Issue by jstebel** [flow123d/flow123d#1556]

**Comment by jstebel** [comment]
Next, the boundary conditions should be fixed as follows:
* In b.c. type `stress` we should prescribe the total stress $\wihdehat\sigma n$ instead of current effective stress: $(\widehat\sigma-\sigma_0)n$. Further, as the initial pressure is already subtracted in $\widehat\sigma$, the value $p_{ref}$, which depends on flow b.c., should be removed. Finally, the sign of the `bc_stress` field should agree with the sign of total stress (positive=tension, negative=compression).
* In b.c. type `traction`, no change is required.
* In b.c. type `displacement` and `displacement_normal`, we wrongly added $\sigma_0 n$, which has to be removed and some test results will necessarily change.

**Comment by Paulie14** [comment]
comparison of older version with master branch on a HM problem of Bukov2 experiment at charon
`pavel_exner/2024_JS_hm_bc_comparison/`

there is a significant difference in pressure at the boundary:
- with master it seems stable (pressure at boundary is as prescribed)
- with older version there is a change of pressure in a boundary layer of elements
- input files differ due to different interface for formula (bparser), otherwise they are equal
- reason for this difference is unknown

**Comment by jbrezmorf** [comment]
I absolutely agree with:
- changes in the bulk equation, i.e. $p_0$ part of initial stress
- consistent sign of BC: $\sigma_T \cdot N = F_{bc} > 0$, regardless of the form of $f_{bc}$.
- prescribing total stress instead of effective stress in BC

I'm not sure about removing the $p_{ref}$. The effect of initial pressure is unrelated to actual boundary pressure in later times.
Consider a Robin-type condition $Q = \alpha_R(t)(p_o -p_i)$ for the flow equation, with $p(x) = p_i$ denoting the trace of the pressure in the domain 
and $p_o$ being the outer pressure. Even if zero explicit outer force is applied, the pressure difference causes total outer force $f_{bc} = (p_i - p_o) N$. If $\alpha_R$ increases from a zero value to some large value, effectively switching from a no-flux condition to a Dirichlet condition, the outer force should drop to nearly zero (zero pressure difference on the boundary) as the pressure difference would be spread through a pressure gradient propagating through the inner domain.

**Comment by jstebel** [comment]
> comparison of older version with master branch on a HM problem of Bukov2 experiment at charon `pavel_exner/2024_JS_hm_bc_comparison/`
> 
> there is a significant difference in pressure at the boundary:
> 
>     * with master it seems stable (pressure at boundary is as prescribed)
> 
>     * with older version there is a change of pressure in a boundary layer of elements
> 
>     * input files differ due to different interface for formula (bparser), otherwise they are equal
> 
>     * reason for this difference is unknown

- The problem in the former version is unrelated to mechanics and to HM coupling. In fact, it can be observed even with Biot alpha=0.
- In the present branch, I also cannot reproduce it.
- However, I was not able to find the source of this strange behaviour and the commit which fixed it.

**Comment by jbrezmorf** [comment]
@Paulie14 and @jstebel, please provide a specific commit for the "older/former version" where the problem is manifested.

**Comment by Paulie14** [comment]
> @Paulie14 and @jstebel, please provide a specific commit for the "older/former version" where the problem is manifested.

edit above; also included in a readme in charon path..

**Comment by jbrezmorf** [comment]
Using a diff between d619691 and  f7bc2896d. 
There are some observations limiting possible cause:
- 439 commits between about 9 bisections
- no change in machanics
- minor change in linsys_PERMON
- some change in `coupling/hm_iterative.cc`:
    -  new implementation of `copy_field`
    - change in making `old_iter_pressure_ptr_` and `old_div_u_ptr_` copies, using FieldFE pointers instead of generic Field API
- no change in flow itself, but new assembly for DarcyFlowOuput, does HM really on the flow output fields, could be that `pressure_p0` is used 
- substantial changes in fields, remove of field_add_potential


There was a change from passing P0 pressure to HM to rather use pressure on edges. I'm not sure if this change is part of the diff. Possibly not.  Clearly 'pressure_p0' was used in the old version d619691. `field_ele_pressure` flow equation field is used in the new version. So these seems to be equivalent.



Could we reproduce the problem in 2D modifying some of the tests?
 We should improve the test set to include non-homogeneous initial pressure and changes in boundary conditions.

**Comment by jstebel** [comment]
Found source of the problem in d619691: User fields do not work in boundary fields.

More details:
* I reduced PE's test to just flow simulation. There is a user field:
```
    user_fields:
          - name: init_p
            scalar_field: !FieldFormula
              value: 100-z
```
which is used in init_pressure and bc_pressure:
```
          - region: BULK
            init_pressure: !FieldFormula
              value: init_p

          - region: .box_outer
            bc_type: dirichlet
            bc_pressure: !FieldFormula
              value: init_p

```

* Replacing the term "init_p" in "bc_pressure" fields by the formula "100-z" fixed the problem.
* On the other hand, init_pressure works OK with the user field.

**Comment by jbrezmorf** [comment]
@jstebel Good work! We will add an appropriate integration test with @dflanderka.

Now the only issue is reference pressure in BC for HM problems.

**Comment by jstebel** [comment]
> I'm not sure about removing the pref. The effect of initial pressure is unrelated to actual boundary pressure in later times. Consider a Robin-type condition Q=αR(t)(po−pi) for the flow equation, with p(x)=pi denoting the trace of the pressure in the domain and po being the outer pressure. Even if zero explicit outer force is applied, the pressure difference causes total outer force fbc=(pi−po)N. If αR increases from a zero value to some large value, effectively switching from a no-flux condition to a Dirichlet condition, the outer force should drop to nearly zero (zero pressure difference on the boundary) as the pressure difference would be spread through a pressure gradient propagating through the inner domain.
> 
> Surely, the meaning of pref must reflect the inclusion of p0 to the initial stress, but I think it still would be necessary in some cases.

Right, that was my mistake. It means we must incorporate the external pressure into mechanical b.c. on hydraulically open boundaries. I see two ways:
a) prescribe either effective or total stress. The user could define bc_effective_stress or bc_total_stress.
b) provide the pressure field in mechanical FieldSet, which would allow to define bc_stress in terms of pressure.
What do you prefer, @jbrezmorf?
